src/io/flutter/run/FlutterRunConfigurationProducer.java

```diff
@@ -6,85 +6,133 @@
+    if (main == null) {
+      return false;
+    }
+
+    config.getRunnerParameters().setFilePath(main.getPath());
+    config.getRunnerParameters()
+      .setWorkingDirectory(FlutterRunnerParameters.suggestDartWorkingDir(context.getProject(), main));
+    config.setGeneratedName();
+
+    final PsiElement elt = sourceElement.get();
+    if (elt != null) {
+      sourceElement.set(elt.getContainingFile());
     }
-    return false;
+    return true;
   }
 
+  /**
+   * Returns true if an existing FlutterRunConfiguration points to the current Dart file.
+   */
   @Override
   public boolean isConfigurationFromContext(final @NotNull FlutterRunConfiguration configuration,
                                             final @NotNull ConfigurationContext context) {
-    final VirtualFile dartFile = getDartFileFromContext(context);
-    return dartFile != null && dartFile.getPath().equals(configuration.getRunnerParameters().getFilePath());
+    final DartFile dart = getDartFile(context);
+    if (dart == null) {
+      return false;
+    }
+
+    final VirtualFile virtual = DartResolveUtil.getRealVirtualFile(dart);
+    return virtual != null && virtual.getPath().equals(configuration.getRunnerParameters().getFilePath());
+  }
+
+  /**
+   * Returns true if Flutter's run configuration should take priority over another one that
+   * applies to the same source file.
+   */
+  @Override
+  public boolean shouldReplace(ConfigurationFromContext self, ConfigurationFromContext other) {
+    // Prefer Flutter runner to plain Dart runner for Flutter code.
+    return DartPlugin.isDartRunConfiguration(other.getConfigurationType());
   }
 
+  /**
+   * Returns the file containing a Flutter app's main() function, or null if not a match.
+   */
   @Nullable
-  private static VirtualFile getRunnableFlutterFileFromContext(final @NotNull ConfigurationContext context) {
-    final PsiElement psiLocation = context.getPsiLocation();
-    final PsiFile psiFile = psiLocation == null ? null : psiLocation.getContainingFile();
-    final VirtualFile virtualFile = DartResolveUtil.getRealVirtualFile(psiFile);
-
-    if (psiFile instanceof DartFile &&
-        virtualFile != null &&
-        ProjectRootManager.getInstance(context.getProject()).getFileIndex().isInContent(virtualFile) &&
-        !DartWritingAccessProvider.isInDartSdkOrDartPackagesFolder(psiFile.getProject(), virtualFile) &&
-        DartResolveUtil.getMainFunction(psiFile) != null &&
-        hasImport((DartFile)psiFile, "package:flutter/material.dart")) { // TODO Determine flutter imports
-      return virtualFile;
+  private static VirtualFile getFlutterEntryFile(final @NotNull ConfigurationContext context) {
+    final DartFile dart = getDartFile(context);
+    if (dart == null) return null;
+
+    if (DartResolveUtil.getMainFunction(dart) == null) return null;
+    if (findImportUrls(dart).noneMatch((url) -> url.startsWith("package:flutter/"))) return null;
```

Comment:
If this becomes more generally useful we may want to extract this test into a method but we can cross that bridge when we get to it.